pubtools/_impl/tracing.py

```diff
@@ -0,0 +1,136 @@
+"""Instrument functions.
+
+Usage:
+    @instrument_func()
+      def func():
+          pass
+
+"""
+import functools
+import logging
+import os
+import threading
+
+from opentelemetry import baggage, context, trace
+from opentelemetry.baggage.propagation import W3CBaggagePropagator
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import (BatchSpanProcessor,
+                                            ConsoleSpanExporter)
+from opentelemetry.trace import Status, StatusCode
+from opentelemetry.trace.propagation.tracecontext import \
+    TraceContextTextMapPropagator
+
+from pubtools.pluggy import pm
+
+propagator = TraceContextTextMapPropagator()
+baggage_propagator = W3CBaggagePropagator()
+log = logging.getLogger(__name__)
+
+
+class TracingWrapper:
+    """Wrapper class to initialize opentelemetry instrumentation"""
+
+    __instance = None
+
+    def __new__(cls):
+        """Create a new instance if one does not exist."""
+        if not os.getenv("OTEL_TRACING", "").lower() == "true":
+            return None
+
+        if TracingWrapper.__instance is None:
+            log.info("Creating TracingWrapper instance")
+            cls.__instance = super().__new__(cls)
+            exporter = (
+                pm.hook.otel_exporter()
+                if pm.hook.otel_exporter()
+                else ConsoleSpanExporter()
```

Comment:
Because I've never run this, I have no idea what the console exporter does.  Have you tried it yet?  Does it seem like a useful default when running locally?

Reply:
Yes, I tried it in my local env. This exporter can print out all spans in the trace in the console, for example:

`
{
    "name": "test_span",
    "context": {
        "trace_id": "0x57eabe45b97e63ba6d734f1b9b751daf",
        "span_id": "0xcb886a22023940bc",
        "trace_state": "[]"
    },
    "kind": "SpanKind.INTERNAL",
    "parent_id": null,
    "start_time": "2023-12-05T01:40:50.679578Z",
    "end_time": "2023-12-05T01:40:50.679598Z",
    "status": {
        "status_code": "UNSET"
    },
    "attributes": {},
    "events": [],
    "links": [],
    "resource": {
        "attributes": {
            "service.name": "client-local-test"
        },
        "schema_url": ""
    }
}
`
The benefit of console exporter is easy to setup.  But, If we'd like to see the relationships between spans clearly, it's not straightforward.
 
For now, I am more using OLTP exporter to send trace data a local [jaeger](https://www.jaegertracing.io/) which provides web UI to see the trace in my local test env.